### DISCUSSION TITLE
Add multiple backups for one cluster

### DISCUSSION THREAD

**Discussion by mahdiG** [cloudnative-pg/cloudnative-pg#2184] [Q&A]
Hi. I need to backup one database cluster into multiple s3 storages. I think backup configuration should accept a list instead of a single item.

**Comment by itay-grudev** [comment]
> Amazon S3 Standard, S3 Standard–IA, S3 Intelligent-Tiering, S3 One Zone-IA, S3 Glacier Instant Retrieval, S3 Glacier Flexible Retrieval, and S3 Glacier Deep Archive are all designed to provide 99.999999999% (11 9's) of data durability of objects over a given year.

Tripling that only increases cost, while the probability decrease of downtime is negligible. 

Furthermore AWS provides a feature known as [Object Lock](https://aws.amazon.com/s3/features/object-lock/) that allows you to prevent deletion or overwriting for a specified retention period. This allows you to ensure that data won't be tampered and you are safe in the event of intentional/accidental deletions and even in the event of a ransomware attack.

Realistically you shouldn't need multiple AWS S3 storages.

That being said, if you are using an S3-like service like MinIO - that is a different conversation. But I still believe that solving it from CNPG is not ideal, for several reasons:

1. There is an additional compute and memory cost with a secondary replication lowering performance.
2. There is an additional bandwidth cost on your cluster also lowering its performance.
3. And last but not least backup high availability probably falls beyond the scope of the operator.

As a final thought you can technically already do this in CNPG, by creating a replica cluster with its own separate backup storage. What's more with this method you can use the original S3 for replication source, removing the two performance limitations as a potential issue.

See here for more details: https://cloudnative-pg.io/documentation/1.18/replica_cluster/

**Comment by mahdiG** [comment]
> That being said, if you are using an S3-like service like MinIO - that is a different conversation.
Yes this is what I meant. For example a user might have S3 from aws, another S3 from another provider, and a self-hosted S3 and want to backup a pg cluster into all of them. It's better to backup into multiple places than only one.

I don't know if this limitation is only for s3 or it's the same story for doing backup for example into S3 and azure at the same time for example. If so, that's a good feature too. But my problem is for multiple S3 for now.

Regarding the reasons you provided, I must say usually you do backup in low load hours so performance/bandwidth issues won't be that much of a problem :) And I think backup/restore is the one of the most important parts of a database. If you have more/better backups, you sleep with less stress :)

Thank you for the provided workaround but if I want to create 3 new clusters cloned from the original cluster and then do 3 backups, that would cost same/more bandwidth but even more performance I believe. But if there's one backup pushed into three S3, it would cost less performance than creating 3 new clusters and then doing backups.

EDIT: 
#2164 This could be another workaround too. If I can add backup configurations in Backup crds, it works. I'll just create multiple backup crds with the s3 configs that I want.

**Comment by itay-grudev** [comment]
You do understand that these backups are not just done once per day, but it's a continuous process. A base backup is performed at whatever schedule you specify, but Write-Ahead Logs are continuously archived all the time.

When restoring a backup you apply the base backups, then you replay all the WAL logs created during the backup process and after it to the Point In Time Recovery moment you selected.

So this is a continuous process constantly using resources. The more write queries you have - the WAL logs you create and need to archive - increasing the compute/memory/bandwidth cost, exactly when you need it.

**Comment by sxd** [comment]
Hi @mahdiG 

Just like @itay-grudev was explaining, the process it's a continuous backup, probably what you can do it's to replicate your Object Storage to three different regions instead of just one?

Turning this into a discussion since it doesn't seem to be an issue, not for now at least

Regards,
